apps/api/src/api/states.ts

```diff
@@ -0,0 +1,292 @@
+	private draftEditTransition = transition(DRAFT, edit, DRAFT, this._onEdit);
+	private draftCloseTransition = transition(DRAFT, close, CLOSED, this._onClose);
+
+	// Rep Review
+	private repReviewCloseTransition = transition(INSTITUTIONAL_REP_REVIEW, close, CLOSED, this._onClose);
+	private repReviewEditTransition = transition(INSTITUTIONAL_REP_REVIEW, edit, DRAFT, this._onEdit);
+	private repReviewRevisionTransition = transition(
+		INSTITUTIONAL_REP_REVIEW,
+		revision_request,
+		REP_REVISION,
+		this._onRevision,
+	);
+	private repReviewSubmitTransition = transition(INSTITUTIONAL_REP_REVIEW, submit, DAC_REVIEW, this._onSubmit);
+
+	// Rep Revision
+	private repRevisionSubmitTransition = transition(REP_REVISION, submit, INSTITUTIONAL_REP_REVIEW, this._onSubmit);
+
+	// DAC Review
+	private dacReviewApproveTransition = transition(DAC_REVIEW, approve, APPROVED, this._onApproved);
+	private dacReviewCloseTransition = transition(DAC_REVIEW, close, CLOSED, this._onClose);
+	private dacReviewEditTransition = transition(DAC_REVIEW, edit, DRAFT, this._onEdit);
+	private dacReviewRevisionTransition = transition(
+		DAC_REVIEW,
+		revision_request,
+		DAC_REVISIONS_REQUESTED,
+		this._onRevision,
+	);
+	private dacReviewRejectTransition = transition(DAC_REVIEW, reject, REJECTED, this._onReject);
+
+	// DAC Revision
+	private dacRevisionSubmitTransition = transition(DAC_REVISIONS_REQUESTED, submit, DAC_REVIEW, this._onSubmit);
+
+	// Revoke Approval
+	private approvalRevokedTransition = transition(APPROVED, revoked, REVOKED, this._onRevoked);
+
+	// All Transitions
+	private applicationTransitions: ApplicationTransitions[] = [
+		this.approvalRevokedTransition,
+		this.dacReviewApproveTransition,
+		this.dacReviewCloseTransition,
+		this.dacReviewEditTransition,
+		this.dacReviewRevisionTransition,
+		this.dacReviewRejectTransition,
+		this.dacRevisionSubmitTransition,
+		this.draftCloseTransition,
+		this.draftEditTransition,
+		this.draftSubmitTransition,
+		this.repReviewCloseTransition,
+		this.repReviewEditTransition,
+		this.repReviewRevisionTransition,
+		this.repReviewSubmitTransition,
+		this.repRevisionSubmitTransition,
+	];
+
+	constructor(application: typeof applications.$inferSelect) {
+		const { id, state } = application;
+		super(state);
+		this._id = id;
+		this._initState = state;
+		this._application = application;
+
+		this.addTransitions(this.applicationTransitions);
+	}
+
+	get id() {
+		return this._id;
+	}
+
+	get state() {
+		return this.getState();
+	}
+}
+
+export const createApplicationStateManager = async ({ id }: { id: number }) => {
+	const database = getDbInstance();
+	const service: ReturnType<typeof applicationService> = applicationService(database);
+
+	const result = await service.getApplicationById({ id });
+	if (!result.success) {
+		return result;
+	}
+
+	const dbRecord = result.data;
+	const appStateManager = new ApplicationStateManager(dbRecord);
+
+	return success(appStateManager);
+};
```

Comment:
Could we instead change this function to take the application as the input? A pattern I expect to see is that an application is read from the DB, some action is taken from it, and then a state transition is required. So would at that time want to create the state manager from the application and perform the action.

Reply:
Alternately, we could have both `createApplicationStateManager(application)` and `getApplicationWithStateManager(id: number)` as a convenience to do both steps at once.

Reply:
Yes, in review the service and result logic should just live in the API ... but then this function just becomes `	const appStateManager = new ApplicationStateManager(dbRecord);`

So `createApplicationStateManager` basically is a convenience method, the class constructor takes the application as an argument